File: prover-android/app/src/main/java/org/nachweis/prover/crypto/Jwe.kt
package org.nachweis.prover.crypto

import org.json.JSONObject
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * JWE compact serialisation, alg ECDH-ES (direct key agreement), enc A128GCM
 * or A256GCM: what the EUDI wallet produces for `direct_post.jwt` (RFC 7516,
 * RFC 7518 sections 4.6 and 5.3).
 */
object Jwe {
    data class Decrypted(val header: JSONObject, val plaintext: ByteArray)

    fun decryptCompact(jwe: String, privateKey: PrivateKey): Decrypted {
        val parts = jwe.trim().split(".")
        require(parts.size == 5) { "JWE compact serialisation needs 5 parts, got ${parts.size}" }
        val (protectedB64, encryptedKeyB64, ivB64, ciphertextB64, tagB64) = parts
        val header = JSONObject(String(b64(protectedB64), Charsets.UTF_8))
        val alg = header.optString("alg")
        val enc = header.optString("enc")
        require(alg == "ECDH-ES") { "unsupported alg $alg (expected ECDH-ES)" }
        require(encryptedKeyB64.isEmpty()) { "ECDH-ES direct mode has an empty encrypted key" }
        val keyBits = when (enc) {
            "A128GCM" -> 128
            "A256GCM" -> 256
            else -> throw IllegalArgumentException("unsupported enc $enc")
        }
        val epk = EcKeys.publicKeyFromJwk(header.getJSONObject("epk"))
        val z = ecdh(privateKey, epk)
        val apu = header.optString("apu", "").let { if (it.isEmpty()) ByteArray(0) else b64(it) }
        val apv = header.optString("apv", "").let { if (it.isEmpty()) ByteArray(0) else b64(it) }
        val cek = concatKdf(z, enc, apu, apv, keyBits)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(cek, "AES"), GCMParameterSpec(128, b64(ivB64)))
        cipher.updateAAD(protectedB64.toByteArray(Charsets.US_ASCII))
        val plaintext = cipher.doFinal(b64(ciphertextB64) + b64(tagB64))
        return Decrypted(header, plaintext)
    }

    fun ecdh(privateKey: PrivateKey, publicKey: PublicKey): ByteArray {
        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(privateKey)
        ka.doPhase(publicKey, true)
        return ka.generateSecret()
    }

    /**
     * NIST SP 800-56A Concat KDF with SHA-256, one round (keys up to 256 bits):
     * H(0x00000001 || Z || len(alg)||alg || len(apu)||apu || len(apv)||apv || keydatalen).
     */
    fun concatKdf(z: ByteArray, algorithmId: String, apu: ByteArray, apv: ByteArray, keyBits: Int): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(ByteBuffer.allocate(4).putInt(1).array())
        md.update(z)
        md.update(lengthPrefixed(algorithmId.toByteArray(Charsets.US_ASCII)))
        md.update(lengthPrefixed(apu))
        md.update(lengthPrefixed(apv))
        md.update(ByteBuffer.allocate(4).putInt(keyBits).array())
        return md.digest().copyOf(keyBits / 8)
    }

    private fun lengthPrefixed(b: ByteArray): ByteArray = ByteBuffer.allocate(4).putInt(b.size).array() + b
    private fun b64(s: String): ByteArray = Base64.getUrlDecoder().decode(s)
}

File: spikes/browser-prover/src/jwe.ts
// Compact JWE decryption with plain WebCrypto, the wallet response path (OpenID4VP direct_post.jwt):
// ECDH-ES direct key agreement (P-256) with the Concat KDF of RFC 7518 section 4.6.2, then A128GCM
// or A256GCM with the protected header as AAD. No library; this is what the browser can do natively.

const enc = new TextEncoder();
type Bytes = Uint8Array<ArrayBuffer>;
const b64uToBytes = (s: string): Bytes => {
  const b = atob(s.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat((4 - (s.length % 4)) % 4));
  return Uint8Array.from(b, (c) => c.charCodeAt(0));
};
const u32be = (n: number): Bytes => new Uint8Array([(n >>> 24) & 255, (n >>> 16) & 255, (n >>> 8) & 255, n & 255]);
const lenPrefixed = (b: Bytes): Bytes => concat(u32be(b.length), b);
const concat = (...parts: Bytes[]): Bytes => {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let o = 0;
  for (const p of parts) {
    out.set(p, o);
    o += p.length;
  }
  return out;
};

export interface Jwk {
  kty: string;
  crv: string;
  x: string;
  y: string;
  d?: string;
}

export async function decryptCompactJwe(jwe: string, privateJwk: Jwk): Promise<{ plaintext: Uint8Array; header: Record<string, unknown> }> {
  const parts = jwe.split(".");
  if (parts.length !== 5) throw new Error(`compact JWE has ${parts.length} parts, expected 5`);
  const [hB64, encKeyB64, ivB64, ctB64, tagB64] = parts;
  const header = JSON.parse(new TextDecoder().decode(b64uToBytes(hB64))) as Record<string, any>;
  if (header.alg !== "ECDH-ES") throw new Error(`alg ${header.alg}, expected ECDH-ES`);
  if (encKeyB64 !== "") throw new Error("ECDH-ES direct mode carries no encrypted key");
  const keyBits = header.enc === "A128GCM" ? 128 : header.enc === "A256GCM" ? 256 : 0;
  if (!keyBits) throw new Error(`enc ${header.enc}, expected A128GCM or A256GCM`);
  const epk = header.epk as Jwk;
  if (!epk || epk.kty !== "EC" || epk.crv !== "P-256") throw new Error("epk is not a P-256 key");

  const subtle = crypto.subtle;
  const priv = await subtle.importKey("jwk", { ...privateJwk, key_ops: ["deriveBits"] } as JsonWebKey, { name: "ECDH", namedCurve: "P-256" }, false, ["deriveBits"]);
  const pub = await subtle.importKey("jwk", { kty: "EC", crv: "P-256", x: epk.x, y: epk.y } as JsonWebKey, { name: "ECDH", namedCurve: "P-256" }, false, []);
  const z = new Uint8Array(await subtle.deriveBits({ name: "ECDH", public: pub }, priv, 256));

  // Concat KDF, one SHA-256 round is enough for 128 or 256 bit keys.
  const otherInfo = concat(
    lenPrefixed(enc.encode(header.enc)),
    lenPrefixed(header.apu ? b64uToBytes(header.apu) : new Uint8Array()),
    lenPrefixed(header.apv ? b64uToBytes(header.apv) : new Uint8Array()),
    u32be(keyBits),
  );
  const derived = new Uint8Array(await subtle.digest("SHA-256", concat(u32be(1), z, otherInfo))).slice(0, keyBits / 8) as Bytes;
  const cek = await subtle.importKey("raw", derived, { name: "AES-GCM" }, false, ["decrypt"]);
  const ciphertext = concat(b64uToBytes(ctB64), b64uToBytes(tagB64));
  const plaintext = new Uint8Array(
    await subtle.decrypt({ name: "AES-GCM", iv: b64uToBytes(ivB64), additionalData: enc.encode(hB64), tagLength: 128 }, cek, ciphertext),
  );
  return { plaintext, header };
}
